runtime/builtin_types.go

```diff
@@ -248,6 +249,82 @@ func builtinMapFn(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return NewList(result...).ToObject(), nil
 }
 
+func builtinFilter(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc < 2 {
+		return nil, f.RaiseType(TypeErrorType, fmt.Sprintf("filter expected 2 arguments, got %d", argc))
+	}
+	fn := args[0]
+	l := args[1]
+	switch {
+	case l.isInstance(TupleType):
+		result := make([]*Object, 0)
+		for _, item := range toTupleUnsafe(l).elems {
+			if fn == None {
+				if ret, raised := IsTrue(f, item); raised != nil {
+					return nil, raised
+				} else if ret {
+					result = append(result, item)
+				}
+			} else {
+				ret, raised := args[0].Call(f, []*Object{item}, nil)
+				if raised != nil {
+					return nil, raised
+				}
+				if ret, raised := IsTrue(f, ret); raised != nil {
+					return nil, raised
+				} else if ret {
+					result = append(result, item)
+				}
+			}
+		}
+		return NewTuple(result...).ToObject(), nil
+	case l.isInstance(StrType):
+		if fn == None {
+			return l, nil
+		}
+		var result bytes.Buffer
+		for _, item := range toStrUnsafe(l).Value() {
+			ret, raised := fn.Call(f, []*Object{NewStr(string(item)).ToObject()}, nil)
+			if raised != nil {
+				return nil, raised
+			}
+			if ret, raised := IsTrue(f, ret); raised != nil {
+				return nil, raised
+			} else if ret {
+				result.WriteRune(item)
+			}
+		}
+		return NewStr(result.String()).ToObject(), nil
+	default:
+		result := make([]*Object, 0)
+		raised := seqForEach(f, l, func(item *Object) (raised *BaseException) {
+			if fn == None {
```

Comment:
Given you do this check a few times I wonder if you should do something like this at the top:

```
filterFunc := IsTrue
if fn != None {
	filterFunc = func(f *Frame, o *Object) (bool, *BaseException) {
		result, raised := fn.Call(f, Args{o}, nil)
		if raised != nil {
			return false, raised
		}
		return IsTrue(f, result)
	}
}
```

And then use filterFunc throughout.